escape double quotes in env variables under windows

schribl opened the pull request:
**Problem description:**
For Windows there is a problem with Javas ProcessBuilder and arguments that contain spaces and contain double quotes as ProcessBuilder adds double quotes around such argument.
See e.g. https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6511002 and https://stackoverflow.com/questions/12124935/processbuilder-adds-extra-quotes-to-command-line

**Solution:**
For Windows all environment variables (only values) are escaped (currently only escapes quotes). No changes made to the Linux (and others) code as I could only reproduce this behavior with a Windows node.

**Notes:**
- This is nearly the same as #215 but does not add additional quotes around every environment variable value and also applies this do Windows only.
- I build this changes locally and tested it in some Jenkins test instance and for my minimal testcases it resolved the issue and did not introduce new issues.

car-roll commented:
bouncing to run against updated CI

schribl commented:
Can I help the process with getting this in, like providing more details/documentation or adding tests for such special characters in environment variables?

car-roll commented:
Hi @schribl if you could add some tests for the  special characters that would be great!

schribl commented:
HI @car-roll I tried implementing some tests for this change (escaping arguments for Windows). But a real test would require the CI to have windows nodes with docker installed IMO.

Doing some other tests require changes of the `DockerClient` design and will test implementation details and not the real function.

So I do not think this can really be tested without changes to the infrastructure or at least I do not see a good way for doing so.

car-roll commented:
Hi @schribl I think what you have is fine as is then. I am curious, do you know if the changes in #215 solve your issue?

car-roll commented:
So the biggest concern is if this change breaks some other use cases. If you look at this method https://github.com/jenkinsci/jenkins/blob/449c5aced523a6e66fe3d6a804e5dbfd5c5c67c6/core/src/main/java/hudson/os/WindowsUtil.java#L53 you can see a bit of checking that goes on here. It might be possible to replace `escapeQuotes` with this `WindowsUtil.quoteArgument` method and see how it handles your local test cases. At the very least, we probably shouldn't do anything if there are no spaces in the argument.  

Would you happen to have a few examples that this PR is addressing?

Thanks @dwnusbaum for the link to the code and basically this response.

schribl commented:
@car-roll @dwnusbaum Thank you! The link indeed looks good and I can understand the concern to break something else with this change.

I will do some more tests and provide the feedback including the results for `WindowsUtil.quoteArgument`. When I did the change back then I think I tried #215 but it did not work as the change was in `DockerClient` and not `WindowsDockerClient` and did not work for this reason (I remember trying both build with some test Jenkins setup).

A simple example I tested with was:
```
withEnv(['TEST="test test2"']) {
    println(env.TEST)
    docker.image("mcr.microsoft.com/dotnet/framework/runtime:4.8").inside() {
        println(env.TEST)
    }
}
```

schribl commented:
I did some more tests and can now confirm that #215 did not resolve the issue as it does change `DockerClient` and not `WindowsDockerClient`. With the link to `WindowsUtil` I added the `.java` file to the `src/main/java/hudson/os` folder and changed the code to:

```java
if (variable.getValue().contains(" ") {
    argb.addMasked(WindowsUtil.quoteArgument(variable.getKey() + "=" + variable.getValue()));
} else {
    argb.addMasked(variable.getKey() + "=" + variable.getValue());
}
```

This also worked for my simple test case but only quoting the value did not work as the function adds quotes around the string and as a result has to quote the full argument. I also think the `if` condition should also check if the key contains a space but this I did not test yet.

@car-roll Is there a better way then copying the `WindowsUtil.java` to the plugin itself?

car-roll commented:
@schribl yes, in this case you will have to bump the jenkins version from 2.176.4 to 2.190.1. `WindowsUtil` was not introduced until after 2.176.4. Adding the space for the `if` condition sounds good as well.

schribl commented:
Thanks @car-roll I did the changes for the proper update and it is using `WindowUtil` now if there are spaces in either key or value of the variable.

dwnusbaum commented on src/main/java/org/jenkinsci/plugins/docker/workflow/client/WindowsDockerClient.java:
```diff
@@ -46,7 +47,11 @@ public String run(@Nonnull EnvVars launchEnv, @Nonnull String image, @CheckForNu
         for (Map.Entry<String, String> variable : containerEnv.entrySet()) {
             argb.add("-e");
-            argb.addMasked(variable.getKey()+"="+variable.getValue());
+            if (variable.getKey().contains("") || variable.getValue().contains(" ")) {
```
Right now this branch always runs, because `String.contains("")` is always true. I think you meant something like this:

```suggestion
            if (variable.getKey().contains("\"") || variable.getValue().contains(" ")) {
```

schribl replied:
It was intended to search for spaces in the the key, but as you pointed out this is already handled by the `WindowsUtil` function

dwnusbaum commented on src/main/java/org/jenkinsci/plugins/docker/workflow/client/WindowsDockerClient.java:
```diff
@@ -46,7 +47,11 @@ public String run(@Nonnull EnvVars launchEnv, @Nonnull String image, @CheckForNu
         for (Map.Entry<String, String> variable : containerEnv.entrySet()) {
             argb.add("-e");
-            argb.addMasked(variable.getKey()+"="+variable.getValue());
+            if (variable.getKey().contains("") || variable.getValue().contains(" ")) {
+                argb.addMasked(WindowsUtil.quoteArgument(variable.getKey() + "=" + variable.getValue()));
+            } else {
+                argb.addMasked(variable.getKey() + "=" + variable.getValue());
+            }
```
That said, `WindowsUtil.quoteArgument` already [checks for a space or quote](https://github.com/jenkinsci/jenkins/blob/449c5aced523a6e66fe3d6a804e5dbfd5c5c67c6/core/src/main/java/hudson/os/WindowsUtil.java#L54) before doing anything, so I think this can be simplified:

```suggestion
            argb.addMasked(WindowsUtil.quoteArgument(variable.getKey() + "=" + variable.getValue()));
```

schribl replied:
Yes, I missed that. Thanks!

schribl commented:
@dwnusbaum Can you check the changes I made regarding your review comments? Thank you!

schribl commented:
@car-roll and @dwnusbaum Friendly reminder on this PR :)
